[FLINK-17547] Support unaligned checkpoints for records spilled to files

rkhachatryan opened the pull request:
## What is the purpose of the change

Implement `SpillingAdaptiveSpanningRecordDeserializer.getUnconsumedBuffer` for records spilled to file (currently, `UnsupportedOperationException` is thrown).


## Brief change log

  - *refactor `SpillingAdaptiveSpanningRecordDeserializer`; extract `SpanningWrapper`*
  - *extract `RefCountingFile` and move it to `flink-core`*
  - *implement `SpanningWrapper.getUnconsumedBuffer` using `RefCountingFile`*

## Verifying this change

  - Added unit tests:  `SpanningWrapperTest`, `MemorySegmentFactoryTest`, `CloseableIteratorTest`
 - Existing integration test: `ChannelPersistenceITCase`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/serialization/SpanningWrapper.java:
```diff
@@ -82,187 +94,171 @@ public SpanningWrapper(String[] tempDirs) {
-				this.spillFileReader = new DataInputViewStreamWrapper(inStream);
-			}
+	private Tuple2<Integer, Integer> readLength(MemorySegment segment, int segmentPosition, int segmentRemaining) throws IOException {
+		int numBytes = min(lengthBuffer.remaining(), segmentRemaining);
```
nit: `numBytes` has a very different meaning on the call stack in `addNextChunkFromMemorySegment()` method, so naming (collision?) is a bit unfortunate.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/serialization/SpanningWrapper.java:
```diff
@@ -82,187 +94,171 @@ public SpanningWrapper(String[] tempDirs) {
-		int toCopy = Math.min(needed, segmentRemaining);
+		int toCopy = min(recordLength - accumulatedRecordBytes, remaining);
+		if (toCopy > 0) {
+			readData(segment, startPos, toCopy);
```
nit: `readData` -> `copyFromSegment` 
`readIntoBuffer` -> `copyIntoBuffer` / `copyFromSegmentIntoBuffer`
`readIntoFile` -> `copyIntoFile` / `copyFromSegmentIntoFile`
?
It wasn't obvious to me what's the direction and purpose of the plain "readData". (Inside `readData` it was already obvious)

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriteRequest.java:
```diff
@@ -46,8 +48,27 @@ static CheckpointInProgressRequest completeOutput(long checkpointId) {
+					iterator.close();
+				} catch (Exception e) {
+					final RuntimeException re = new RuntimeException("unable to close iterator", e);
+					re.addSuppressed(throwable);
```
It depends on which exception is thrown afterward. Currently, it is `re` to which `throwable` is added as suppressed.

Do you think we should throw `throwable` instead?

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/serialization/SpanningWrapper.java:
```diff
@@ -183,22 +196,26 @@ private void updateLength(int length) throws IOException {
+		if (spillingChannel != null) {
+			spillingChannel.force(false);
+		}
+		return CloseableIterator.flatten(
+			toSingleBufferIterator(wrapInt(recordLength)),
+			new FileBasedBufferIterator(spillFile, min(accumulatedRecordBytes, recordLength), fileBufferSize),
+			leftOverData == null ? empty() : toSingleBufferIterator(wrapCopy(leftOverData.getArray(), leftOverStart, leftOverLimit))
+		);
```
It flattens 3 iterators, one of which can be empty. I tried to put it all into `FileBasedBufferIterator` but it doesn't simplify anything (as there will be many checks there) and makes it more error-prone.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/serialization/SpanningWrapper.java:
```diff
@@ -82,187 +94,171 @@ public SpanningWrapper(String[] tempDirs) {
+		int numBytes = min(lengthBuffer.remaining(), segmentRemaining);
+		segment.get(segmentPosition, lengthBuffer, numBytes);
+		if (lengthBuffer.hasRemaining()) {
+			return POSITION_AND_SIZE_IF_LENGTH_NOT_FULLY_READ;
```
To avoid unnecessary creation of the same `Tuple2` each time (and make it more readable).

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/serialization/SpanningWrapper.java:
```diff
@@ -43,15 +46,18 @@
 
-	private static final int THRESHOLD_FOR_SPILLING = 5 * 1024 * 1024; // 5 MiBytes
-	private static final int FILE_BUFFER_SIZE = 2 * 1024 * 1024;
+	private static final int DEFAULT_THRESHOLD_FOR_SPILLING = 5 * 1024 * 1024; // 5 MiBytes
+	private static final int DEFAULT_FILE_BUFFER_SIZE = 2 * 1024 * 1024;
```
I agree that it could be less but it's the same constant used for "normal" reading of "spill" file.
I don't think it should be a different number of that we should change it in this PR (and without some experimentation).

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/serialization/SpillingAdaptiveSpanningRecordDeserializer.java:
```diff
@@ -158,7 +113,10 @@ public void clear() {
 
+	interface BufferWrapper {
+		Optional<MemorySegment> getUnconsumedSegment() throws IOException;
+	}
```
It is implemented by both wrappers (used in `SpillingAdaptiveSpanningRecordDeserializer`).
I think they could further be unified (unless we drop `NonSpanningWrapper` and merge 
`SpillingAdaptiveSpanningRecordDeserializer` with `SpanningWrapper`).

pnowojski replied:
I've seen it was implemented, but the interface itself was not being used anywhere, so I'm not sure if it has much value.

rkhachatryan replied:
Removed.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/serialization/NonSpanningWrapper.java:
```diff
@@ -47,27 +64,33 @@ void clear() {
+			segment.copyTo(position, target, 0, remaining());
+			return Optional.of(target);
+		} else {
 			return Optional.empty();
```
The other point of view is that `if` shouldn't negate the condition :)
Will swap it.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/serialization/NonSpanningWrapper.java:
```diff
@@ -290,4 +313,55 @@ public int read(byte[] b, int off, int len) {
+		return Tuple2.of(segment.wrap(position, numBytes), numBytes);
+	}
+
+	int copyTo(byte[] dst) {
+		final int numBytesChunk = remaining();
+		segment.get(position, dst, 0, numBytesChunk);
+		return numBytesChunk;
+	}
+
+	/**
+	 * We have an incomplete length add our part of the length to the length buffer.
+	 */
+	void transferTo(ByteBuffer dst) {
+		segment.get(position, dst, remaining());
+		clear();
+	}
```
`copyTo` only copies data, while `transferTo` transfers "ownership" (and that's why it calls `clear()`). Added Javadoc for this method.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/serialization/SpillingAdaptiveSpanningRecordDeserializer.java:
```diff
@@ -91,62 +86,22 @@ public DeserializationResult getNextRecord(T target) throws IOException {
-		if (this.spanningWrapper.hasFullRecord()) {
-			// get the full record
-			target.read(this.spanningWrapper.getInputView());
+		} else if (nonSpanningWrapper.hasRemaining()) {
```
It would change the existing logic and make less readable IMO.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/serialization/SpillingAdaptiveSpanningRecordDeserializer.java:
```diff
@@ -91,62 +86,22 @@ public DeserializationResult getNextRecord(T target) throws IOException {
-			this.nonSpanningWrapper.clear();
-			return DeserializationResult.PARTIAL_RECORD;
-		}
+		if (nonSpanningWrapper.canReadLength()) {
```
Renamed as you suggested below.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/serialization/SpanningWrapper.java:
```diff
@@ -82,187 +94,171 @@ public SpanningWrapper(String[] tempDirs) {
-				}
-			}
+		Tuple2<Integer, Integer> positionAndRemaining = readLengthIfNeeded(segment, offset, numBytes);
+		int startPos = positionAndRemaining.f0;
```
`startPosition`  :-1:, but ok :)


> those things are the same

They are not: we need to use old values to compute new `leftOverLimit`.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriteRequest.java:
```diff
@@ -46,8 +47,23 @@ static CheckpointInProgressRequest completeOutput(long checkpointId) {
+						checkArgument(buffer.isBuffer());
+						writer.writeInput(info, buffer);
+					}
+				} finally {
+					iterator.close();
+				}
+			},
+			throwable -> iterator.close(),
```
are you sure this won't be called twice? Or if it will, that it will not double release the buffers/files?

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriteRequestExecutorImpl.java:
```diff
@@ -67,8 +70,11 @@ void run() {
-			dispatcher.fail(thrown == null ? new CancellationException() : thrown);
+			try {
+				closeAll(this::cleanupRequests, () -> dispatcher.fail(thrown == null ? new CancellationException() : thrown));
+			} catch (Exception e) {
+				LOG.error("unable to terminate properly", e);
+			}
```
we had quite a bit of problems with error logging like that polluting the logs failing e2e tests. It would be better to add `e` as suppressed exception to the `thrown`, like `org.apache.flink.streaming.runtime.tasks.StreamTask#invoke` is doing.
```
} catch (Exception e) {
	dispatcher.fail(ExceptionUtils.firstOrSuppressed(e, thrown));
}
```
?

rkhachatryan replied:
I'll update the `thrown` field (`dispatcher.fail` is already called two lines above).

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/serialization/NonSpanningWrapper.java:
```diff
@@ -290,4 +309,49 @@ public int read(byte[] b, int off, int len) {
+		clear();
+	}
+
+	Tuple2<DeserializationResult, Integer> getNextRecord(IOReadableWritable target) throws IOException {
+		int recordLen = readInt();
+		if (canReadRecord(recordLen)) {
+			return readInto(target);
+		} else {
+			return new Tuple2<>(PARTIAL_RECORD, recordLen);
+		}
+	}
```
nit: by wrapper I meant a POJO not Tuple2, but this is probably fine.

However now I'm a bit worried about performance implications of this - it's a critical/hot path, and we are allocating new object on it.

rkhachatryan replied:
I'll create a POJO and will reuse the instance (it's a single thread).

rkhachatryan commented:
Thanks for the fast response.
I've added two fixup commits to address the issues.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriteRequestExecutorImpl.java:
```diff
@@ -67,8 +70,12 @@ void run() {
 		} finally {
-			cleanupRequests();
-			dispatcher.fail(thrown == null ? new CancellationException() : thrown);
+			try {
+				closeAll(this::cleanupRequests, () -> dispatcher.fail(thrown == null ? new CancellationException() : thrown));
+			} catch (Exception e) {
+				//noinspection NonAtomicOperationOnVolatileField
+				thrown = ExceptionUtils.firstOrSuppressed(e, thrown);
+			}
```
nit: This is a really hacky piece of code :)

I'm fine with keeping it as it is to not trigger another re-build over a minor issue.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/FileBasedBufferIterator.java:
```diff
@@ -85,6 +90,6 @@ private int read(byte[] buffer) {
 	@Override
 	public void close() throws Exception {
-		closeAll(stream, file::release);
+		closeAll(stream, file::release, () -> buffersToClose.forEach(Buffer::recycleBuffer));
```
Why are we recycling buffers at the end, on close and not on the fly? It can mean GBs of extra (unnecessary) memory usage if we keep all unconsumed buffers from different channels. The ownership of the buffers returned from the iterator should belong to the caller of `next()` method.

rkhachatryan replied:
I've changed it to recycle only the last buffer in `close`.
I wanted to avoid extra complexity but it turned out not that complicated.

(the reason why buffers need to be recycled by iterator rather than by client is that for in-memory collections we always want to recycle all the elements, even if iteration wasn't started)

pnowojski replied:
I don't understand it how is it working right now. Where are the remaining (non last) buffers recycled? And why do you have to handle the last buffer differently? As I mentioned before, I think:
> The ownership of the buffers returned from the iterator should belong to the caller of next() method.

so buffers' ownership should be handed over, and owner should be always responsible to recycle it when it's done using it.

rkhachatryan replied:
> I don't understand it how is it working right now. Where are the non last buffers recycled? 

In the beginnning of `FileBasedBufferIterator.next()`.

> And why do you have to handle the last buffer differently?

Because `next()` won't be called anymore.

> The ownership of the buffers returned from the iterator should belong to the caller of next() method.

The client doesn't know whether the iterator is lazy or not. So if failure happens before iteration, it would have to read the whole file just to recycle the buffers.

> I guess we are avoiding memory leaks right now, only because the returned Buffers are not pooled and using simple FreeingBufferRecycler?

No, the buffers are recycled explicitly (see above).

pnowojski replied:
> The client doesn't know whether the iterator is lazy or not. So if failure happens before iteration, it would have to read the whole file just to recycle the buffers.

? Why? As I wrote before, the buffers that were not yet returned, should be recycled by it's current owner - in that case `CloseableIterator`. If the buffers were even not yet created, because it's a lazy iterator, even better - nothing to do in the `close()`. 

> In the beginnning of FileBasedBufferIterator.next().

If buffers were returned from the iterator, iterator is no long it's owner and they shouldn't be recycled by the iterator.

rkhachatryan replied:
I think you are right, these ownership rules are more clear (though implementation will be a bit more complex and less optimal in collections iterators).
I'll update the PR.

rkhachatryan replied:
Done.

rkhachatryan commented:
I've rebased the changes to the latest master and squashed the commits.

Azure [build](https://github.com/rkhachatryan/flink/runs/681261946) in my private account is green.

@flinkbot run azure
